2_process/src/add_dynamic_attr.R

```diff
@@ -0,0 +1,796 @@
+add_dyn_attrs_to_reaches <- function(attrs, dyn_cols, start_date, end_date,
+                                     baseflow = NULL, 
+                                     CAT_Land = NULL, Upstream_Land = NULL, 
+                                     gridMET = NULL,
+                                     attr_prefix, Upstream_Land_prefix = NULL,
+                                     lag_table){
+  #' @description computes dynamic attributes for each reach based on the provided dates
+  #' 
+  #' @param attrs table of static attributes (columns) for each reach (rows)
+  #' @param dyn_cols character vector of keywords to search for columns that 
+  #' should be converted into dynamic attributes.
+  #' @param start_date first date to compute attributes
+  #' @param end_date last date to compute attributes
+  #' @param baseflow table of monthly baseflow estimates for each reach
+  #' @param CAT_Land table of catchment land cover class proportions
+  #' @param Upstream_Land table of total upstream land cover class proportions
+  #' @param gridMET table of gridment data. Row length = dates x PRMS reach
+  #' @param attr_prefix character vector of prefixes for the attributes 
+  #' (e.g., CAT, TOT)
+  #' @param Upstream_Land_prefix character vector of prefixes for the upstream 
+  #' land cover attributes (e.g., TOT, ACC)
+  #' @param lag_table table with 3 columns: attributes, lags, and lag_unit.
+  #' This is used to compute the lagged attributes
+  #' 
+  #' @return tbl with the added features.
+  
+  #make a dynamic features dataframe
+  dates <- seq(as.Date(start_date), as.Date(end_date), 1)
+  segs <- sort(unique(attrs$PRMS_segid))
+  df <- tibble(Date = dates, seg = segs[1])
+  for(i in 2:length(segs)){
+    df_tmp <- tibble(Date = dates, seg = segs[i])
+    df <- rbind(df, df_tmp)
+  }
+  rm(df_tmp, dates, i)
+  
+  #Special handling for each variable
+  if ('HDENS' %in% dyn_cols){
+    #Housing density has the format HDENSYY
+    tmp_attrs <- select(attrs, PRMS_segid, contains('HDENS'))
+    
+    #Convert the column names to 4-digit years
+    col_years <- str_split(colnames(tmp_attrs)[-1], 
+                           pattern = '_', simplify = TRUE)[,2] %>%
+      substr(start = 6, stop = 7) %>%
+      get_four_digit_year()
```

Comment:
This works, but it assumes that the structure of the HDENS names will always be the same. I don't think we'd expect any changes, but nonetheless it might be more robust to replace `col_years` with:

```r
   col_years <- stringr::str_extract(names(tmp_attrs)[-1], "[[:digit:]]+") %>%
      get_four_digit_year()
```